django/db/models/query.py

```diff
@@ -670,11 +670,20 @@ async def acreate(self, **kwargs):
     acreate.alters_data = True
 
     def _prepare_for_bulk_create(self, objs):
+        objs_with_pk, objs_without_pk = [], []
         for obj in objs:
-            if not obj._is_pk_set():
-                # Populate new PK values.
+            if isinstance(obj.pk, DatabaseDefault):
+                objs_without_pk.append(obj)
+            elif obj._is_pk_set():
+                objs_with_pk.append(obj)
+            else:
                 obj.pk = obj._meta.pk.get_pk_value_on_save(obj)
+                if obj._is_pk_set():
+                    objs_with_pk.append(obj)
+                else:
+                    objs_without_pk.append(obj)
```

Comment:
```suggestion
                objs_without_pk.append(obj)
```
I'm not sure if I've missed something but I can change to the following without any bulk_create test failures. Are we missing a test for this or can it be simplified?

Reply:
@sarahboyce through a bit of git-blame if found the test that would fail.

`get_pk_value_on_save` was introduced in 8adc59038cdc6ce4f9170e4de2d716d940e136b3 to [expose a hook for these reasons](https://code.djangoproject.com/ticket/23617#comment:11).

Whether or not this hook is still relevant today now that we support `db_default` is up for discussion but if you run the `model_fields.test_uuid` suite you'll notice that this code is exercised.

Reply:
Thank you Simon :star: